src/openfermion/ops/_symbolic_operator.py

```diff
@@ -0,0 +1,301 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"""SymbolicOperator is a base class for QubitOperator and FermionOperator"""
+import copy
+import itertools
+
+import numpy
+from abc import abstractmethod
+
+
+EQ_TOLERANCE = 1e-12
+
+
+class SymbolicOperatorError(Exception):
+    pass
+
+
+class SymbolicOperator(object):
+    """
+    The base class for QubitOperator and FermionOperator. All methods defined
+    here can be accessed from FermionOperator or QubitOperator objects.
+    This is an abstract class and objects of this type cannot be created,
+    only those of subclasses.
+
+    Subclasses are sums of terms of operators for a particular category of
+    particle. Subclasses support addition and multiplication with objects of
+    the same type.
+
+    Attributes:
+        terms (dict):
+            **key** (tuple of tuples): Each tuple represents a fermion term,
+            i.e. a tensor product of fermion ladder operators with a
+            coefficient. The first element is an integer indicating the
+            mode on which a ladder operator acts and the second element is
+            a bool, either '0' indicating annihilation, or '1' indicating
+            creation in that mode; for example, '2^ 5' is ((2, 1), (5, 0)).
+            **value** (complex float): The coefficient of term represented by
+            key.
+    """
+
+    @abstractmethod
+    def __init__(self):
+        pass
+
+    @abstractmethod
+    def __imul__(self):
+        pass
+
+    @abstractmethod
+    def __str__(self):
+        pass
+
+    @abstractmethod
+    def __repr__(self):
+        pass
+
+    @classmethod
+    def zero(cls):
+        """
+        Returns:
+            additive_identity (SymbolicOperator):
+                A symbolic operator o with the property that o+x = x+o = x for
+                all fermion operators x.
+        """
+        # Maybe throw a TypeError if called using SymbolicOperator? Or maybe
```

Comment:
I think it is okay to let users initialize SymbolicOperator. I don't know why they would, but it seems like a strange trap for them to fall into unless they know what they are doing.